archive/tar_opts_windows.go

```diff
@@ -18,14 +18,24 @@ package archive
 
 import (
 	"context"
+	"fmt"
 	"io"
 
+	"github.com/Microsoft/go-winio"
 	"github.com/Microsoft/hcsshim/pkg/ociwclayer"
 )
 
 // applyWindowsLayer applies a tar stream of an OCI style diff tar of a Windows layer
 // See https://github.com/opencontainers/image-spec/blob/main/layer.md#applying-changesets
 func applyWindowsLayer(ctx context.Context, root string, r io.Reader, options ApplyOptions) (size int64, err error) {
+	// It seems that in certain situations, like having the containerd root and state on a file system hosted on a
+	// mounted VHDX, we need SeSecurityPrivilege when opening a file with winio.ACCESS_SYSTEM_SECURITY. This happens
+	// in the base layer writer in hcsshim when adding a new file.
+	if err := winio.EnableProcessPrivileges([]string{winio.SeSecurityPrivilege}); err != nil {
+		return 0, fmt.Errorf("enabling privileges: %w", err)
+	}
+	defer winio.DisableProcessPrivileges([]string{winio.SeSecurityPrivilege})
+
 	return ociwclayer.ImportLayerFromTar(ctx, r, root, options.Parents)
```

Comment:
Since we're wrapping a single function call, we can probably use `winio.RunWithPrivilege` (which I only just learned about) to make this a thread-local privilege boost.

This is probably true of the other privilege-taking code, which would help with #1681, see https://github.com/search?q=repo%3Acontainerd%2Fcontainerd%201681&type=code, but those privilege-holds have wider scope so should probably not be touched until #8043 lands for test coverage.

Reply:
Hmm, according to https://github.com/moby/moby/pull/28002#discussion_r86263192, `winio.RunWithPrivilege` wasn't enough in the Docker-equivalent of this code-path, probably due to the use of goroutines in hcsshim which end up on a different thread.

Interestingly, if that's the case, it might make sense for hcsshim to take per-thread privileges on goroutines _it_ spawns? The hcsshim API docs say that thread-local privileges should be sufficient though, and I haven't really traced through all the cases.

Reply:
In this case, it seems to work. I updated this PR to use ```RunWithPrivileges()```, and doing a pull of an image seems to work:

```powershell
PS C:\Users\Administrator\work\rebased\go-winio> ctr i pull mcr.microsoft.com/windows/nanoserver:ltsc2022
mcr.microsoft.com/windows/nanoserver:ltsc2022:                                 resolved       |++++++++++++++++++++++++++++++++++++++|
index-sha256:3bb312dcc36fa551766a13cf2fb8e2be90726774589c96bea9198a26307bf2a2: downloading    |++++++++++++++++++++++++++++++++++++++|  350.0 B/350.0 B
mcr.microsoft.com/windows/nanoserver:ltsc2022:                                    resolved       |++++++++++++++++++++++++++++++++++++++|
index-sha256:3bb312dcc36fa551766a13cf2fb8e2be90726774589c96bea9198a26307bf2a2:    done           |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:8e12c98b4d5dd7489f2ced02f88a1e90346d7d6dbb44e161f6c0ef8549d69558: done           |++++++++++++++++++++++++++++++++++++++|
layer-sha256:9e0d440fca83fb7bc10746dfa71673c7187c16abb63a1612e6c3bdfe13786652:    done           |++++++++++++++++++++++++++++++++++++++|
config-sha256:0c7fb865d8dc8e3c00acb2a6a7dc08be3474e9a36ea48ce2407553fc8eb0a434:   done           |++++++++++++++++++++++++++++++++++++++|
elapsed: 9.6 s                                                                    total:  112.9  (11.8 MiB/s)
unpacking windows/amd64 sha256:3bb312dcc36fa551766a13cf2fb8e2be90726774589c96bea9198a26307bf2a2...
done: 7.1672897s
```

And without this change I get:

```powershell
PS C:\Users\Administrator\work\rebased\go-winio> ctr i pull mcr.microsoft.com/windows/nanoserver:ltsc2022
mcr.microsoft.com/windows/nanoserver:ltsc2022:                                 resolved       |++++++++++++++++++++++++++++++++++++++|
index-sha256:3bb312dcc36fa551766a13cf2fb8e2be90726774589c96bea9198a26307bf2a2: downloading    |++++++++++++++++++++++++++++++++++++++|  350.0 B/350.0 B
mcr.microsoft.com/windows/nanoserver:ltsc2022:                                    resolved       |++++++++++++++++++++++++++++++++++++++|
mcr.microsoft.com/windows/nanoserver:ltsc2022:                                    resolved       |++++++++++++++++++++++++++++++++++++++|
index-sha256:3bb312dcc36fa551766a13cf2fb8e2be90726774589c96bea9198a26307bf2a2:    done           |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:8e12c98b4d5dd7489f2ced02f88a1e90346d7d6dbb44e161f6c0ef8549d69558: done           |++++++++++++++++++++++++++++++++++++++|
config-sha256:0c7fb865d8dc8e3c00acb2a6a7dc08be3474e9a36ea48ce2407553fc8eb0a434:   done           |++++++++++++++++++++++++++++++++++++++|
layer-sha256:9e0d440fca83fb7bc10746dfa71673c7187c16abb63a1612e6c3bdfe13786652:    done           |++++++++++++++++++++++++++++++++++++++|
elapsed: 7.2 s                                                                    total:  112.0  (15.6 MiB/s)
unpacking windows/amd64 sha256:3bb312dcc36fa551766a13cf2fb8e2be90726774589c96bea9198a26307bf2a2...
time="2023-03-15T05:36:12-07:00" level=info msg="apply failure, attempting cleanup" error="failed to extract layer sha256:3291301d72b5eb7f585ad7cfd3bf17d2e5ca9a267789d4348e1a6006b1b255fe: Failed to safefile.OpenRelative failed in Win32: open \\\\?\\F:\\containerd\\root\\io.containerd.snapshotter.v1.windows\\snapshots\\1\\Files: Access is denied. (0x5) Files: unknown" key="extract-622345000-RpzD sha256:3291301d72b5eb7f585ad7cfd3bf17d2e5ca9a267789d4348e1a6006b1b255fe"
ctr: failed to extract layer sha256:3291301d72b5eb7f585ad7cfd3bf17d2e5ca9a267789d4348e1a6006b1b255fe: Failed to safefile.OpenRelative failed in Win32: open \\?\F:\containerd\root\io.containerd.snapshotter.v1.windows\snapshots\1\Files: Access is denied. (0x5) Files: unknown
```

Reply:
I converted this to draft for now. Will wait for #8043 to merge before trying to merge this. You are right, the test suite enabled in that PR should be run against this change. Although, to actually test it, we'd need to create a VHDx, format it, mount it and use it as a root.